cmd/kubeadm/app/preflight/checks.go

```diff
@@ -414,6 +417,43 @@ func (hc HostnameCheck) Check() (warnings, errors []error) {
 	return warnings, errors
 }
 
+// HostnameUniqueCheck ensures that the node hostname is unique in the cluster.
+type HostnameUniqueCheck struct {
+	cfg *kubeadmapi.JoinConfiguration
+}
+
+// Name will return HostnameUnique as name for HostnameUniqueCheck.
+func (HostnameUniqueCheck) Name() string {
+	return "HostnameUnique"
+}
+
+// Check validates if node hostname is unique.
+func (huc HostnameUniqueCheck) Check() (warnings, errors []error) {
+	glog.V(1).Infof("checking whether the given hostname is unique in the cluster")
+	errors = []error{}
+	warnings = []error{}
+	hostname := huc.cfg.NodeRegistration.Name
+	tlsBootstrapCfg, err := discovery.For(huc.cfg)
+	if err != nil {
+		warnings = append(warnings, fmt.Errorf("hostname unique check failed: %s", err))
+	}
+	client, err := kubeconfigutil.ToClientSet(tlsBootstrapCfg)
+	if err != nil {
+		warnings = append(warnings, fmt.Errorf("hostname unique check failed: %s", err))
+	}
+	nodes, err := client.CoreV1().Nodes().List(metav1.ListOptions{})
```

Comment:
would it make more sense try to get node with same name, instead of list + iteration?